### DISCUSSION TITLE
Writing custom modules to interface with baresip?

### DISCUSSION THREAD

**Discussion by nonibytes** [baresip/baresip#2932] [Q&A] [answered]
As part of my application, I want to write a custom module that interfaces with baresip and can perform additional actions programmatically when certain events occur. For example doing a database lookup, triggering API endpoints, etc.

Can this be done with baresip?

How can I do this?

I know it's possible to have shell scripts triggered, but what I want to do is beyond shell scripts. I want a tighter coupled access to the baresip instance.

**Comment by aonoa** [comment, marked as answer]
Using curl to access APIs

```
find_package(CURL)
if(NOT CURL_FOUND)
  return()
endif()
```

```
#include <curl/curl.h>

#define URL    "127.0.0.1:8888/file?filename=%s/dump-test%d.pcm"

static CURL *curl = NULL;
```

```
static int module_init(void)
{
        ....
        curl = curl_easy_init();
        if (!curl)
        {
                fprintf(stderr,"curl init failed\n");
                return -1;
        }

        return 0;
}
```
```
static int decode(struct aufilt_dec_st *st, struct auframe *af)
{
        ....
        // send filename
        char url[255]={0};
        (void)re_snprintf(url, sizeof(url),
                          URL,
                          file_path,
                          (sf->num-1)/4);
        curl_easy_setopt(curl,CURLOPT_URL,url);
        curl_easy_perform(curl);
        ....
}
```
```
static int module_close(void)
{
        ....
        curl_easy_cleanup(curl);
        return 0;
}

```

**Reply by nonibytes** [reply]
Thanks for the example!

**Comment by nonibytes** [comment]
For anyone who comes across this post in the future, the re library can be used to make http requests:

https://github.com/baresip/re

Also, it is possible to build a standalone baresip-based application if a using a module isn't feasible.

https://github.com/baresip/baresip/wiki/Using-baresip-as-a-library
